File: src/main/java/com/skocur/imagecipher/tools/UpdateChecker.java
package com.skocur.imagecipher.tools;

import com.skocur.imagecipher.rest.di.DaggerServiceComponent;
import com.skocur.imagecipher.rest.di.ServiceComponent;
import com.skocur.imagecipher.rest.di.ServiceModule;
import com.skocur.imagecipher.rest.github.GitHubService;
import com.skocur.imagecipher.rest.github.Release;
import java.io.IOException;
import java.util.Date;
import java.util.function.Supplier;
import javax.inject.Inject;
import javax.inject.Named;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import retrofit2.Call;
import retrofit2.Response;

public class UpdateChecker {

  @Inject
  @Named("GitHubService")
  public GitHubService gitHubService;

  private static final Logger logger = LogManager.getLogger();

  public UpdateChecker() {
    logger.info("Injecting GitHub Service");
    ServiceComponent serviceComponent = DaggerServiceComponent.builder()
        .serviceModule(new ServiceModule()).build();
    serviceComponent.injectUpdater(this);
  }

  /**
   * @return Boolean - status whether app update was processed or not
   */
  public boolean checkForUpdates(Supplier<Boolean> supplier) {
    logger.info("Checking for updates");
    Call<Release> releaseCall = gitHubService.getLatestRelease();
    try {
      Response<Release> response = releaseCall.execute();
      Release release = response.body();

      if (release == null) {
        logger.error("Release response body is null");
        return false;
      }

      Date buildTime = ManifestReader.getBuildTime();
      if (buildTime == null) {
        logger.warn("Build-Time is null, probably manifest is missing");
        return false;
      }

      if (buildTime.before(release.publishedDate)) {
        return supplier.get();
      } else {
        logger.debug("No new version is available. Current build date: " + buildTime.toString());
      }
    } catch (IOException e) {
      logger.error(e);
    }

    return false;
  }
}

File: app/src/main/kotlin/com/imagecipher/app/tools/UpdateChecker.kt
package com.imagecipher.app.tools

import com.imagecipher.app.rest.github.GitHubService
import com.imagecipher.app.rest.github.Release
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.util.function.Supplier

class UpdateChecker {
    private val gitHubService: GitHubService = Retrofit.Builder()
        .baseUrl("https://api.github.com/")
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(GitHubService::class.java)

    init {
        logger.info("Initializing GitHub Service")
    }

    /**
     * @return Boolean - status whether app update was processed or not
     */
    fun checkForUpdates(supplier: Supplier<Boolean?>): Boolean {
        logger.info("Checking for updates")
        val releaseCall: Call<Release> = gitHubService.getLatestRelease()
        try {
            val response = releaseCall.execute()
            val release = response.body()

            if (release == null) {
                logger.error("Release response body is null")
                return false
            }

            val buildTime = ManifestReader.getBuildTime()
            if (buildTime == null) {
                logger.warn("Build-Time is null, probably manifest is missing")
                return false
            }

            if (buildTime.before(release.publishedDate)) {
                return supplier.get()!!
            } else {
                logger.debug("No new version is available. Current build date: " + buildTime.toString())
            }
        } catch (e: IOException) {
            logger.error(e)
        }

        return false
    }

    companion object {
        private val logger: Logger = LogManager.getLogger()
    }
}
